package devicepackage;

message ColorList {
  repeated Color colors = 1;
}

message Color {
  required int32 r = 1;
  required int32 g = 2;
  required int32 b = 3;
}

message Room {
  repeated DeviceInRoom devicesInRoom = 1;
}

message DeviceInRoom {
  required Settings settings = 1;
  required PositionInRoom position = 2;
}

message PositionInRoom {
  required float x = 1;
  required float y = 2;
  required float z = 3;
  required Plane plane = 4;
  required bool inverted = 5;
}

enum Plane {
  XY = 1;
  XZ = 2;
  YZ = 3;
}

message DisplayPositions {
  repeated Position positions = 1;
}

message Position {
  required float x = 1;
  required float y = 2;
}

message SavedDevices {
  repeated Device devices = 1;
}

message Device {
  required string name = 1;
  required Settings settings = 2;
}

message Settings {
  oneof settings {
    Strip strip = 2;
    Arc arc = 3;
    Grid grid = 4;
    BucketedGrid bucketed_grid = 5;
    BucketedVariableColumns bucketed_variable_columns = 6;
  }
}

message RepeatedNumber {
  repeated int32 repeated_number = 1;
}

enum Unit {
  MM = 0;
  IN = 1;
}

message Dimension {
  required Unit unit = 1;
  required float dimension = 2;
}

message Strip {
  required int32 leds = 1;
  required Dimension spacing = 2;
}

message Arc {
  required int32 leds = 1;
  required Dimension radius = 2;
  oneof dimension {
    Dimension arc_length = 3;
    float angle = 4;
    Dimension chord_length = 5;
  }
}

message Grid {
  required int32 rows = 1;
  required int32 columns = 2;
  required Dimension row_spacing = 3;
  required Dimension column_spacing = 4;
}

message BucketedGrid {
  required int32 rows = 1;
  required int32 columns = 2;
  required Dimension row_spacing = 3;
  required Dimension column_spacing = 4;
  required Settings bucket = 5;
}

message BucketedVariableColumns {
  repeated uint32 row_spacing = 2;
  repeated RepeatedNumber column_spacing = 3;
  required Unit units = 4;
  required Settings bucket = 5;
}